Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,533 @@
+    }
+}
+
+// This is the key we use for merging events in our internal hashtable
+struct fdEventKey: Hashable {
+    var fileDescriptor : CInt
+    var registrationID : UInt16 // we just have the truncated lower 16 bits of the registrationID
+
+    init(_ f: CInt, _ s : UInt16) {
+        self.fileDescriptor = f
+        self.registrationID = s
+    }
+}
+
+
+final internal class Uring {
+    internal static let POLLIN: CUnsignedInt = numericCast(CNIOLinux.POLLIN)
+    internal static let POLLOUT: CUnsignedInt = numericCast(CNIOLinux.POLLOUT)
+    internal static let POLLERR: CUnsignedInt = numericCast(CNIOLinux.POLLERR)
+    internal static let POLLRDHUP: CUnsignedInt = CNIOLinux_POLLRDHUP() // numericCast(CNIOLinux.POLLRDHUP) 
+    internal static let POLLHUP: CUnsignedInt = numericCast(CNIOLinux.POLLHUP)
+    internal static let POLLCANCEL: CUnsignedInt = 0xF0000000 // Poll cancelled, need to reregister for singleshot polls
+
+    private var ring = io_uring()
+    private let ringEntries: CUnsignedInt = 8192
+    private let cqeMaxCount : UInt32 = 8192 // this is the max chunk of CQE we take.
+
+    var cqes : UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>
+    var fdEvents = [fdEventKey : UInt32]() // fd, sequence_identifier : merged event_poll_return
+    var emptyCqe = io_uring_cqe()
+
+    internal func fd() -> Int32 {
+       return ring.ring_fd
+    }
+
+    func _dumpCqes(_ header:String, count: Int = 1) {
+        #if SWIFTNIO_IO_URING_DEBUG_DUMP_CQE
+        func _debugPrintCQE(_ s: String) {
+            print("Q [\(NIOThread.current)] " + s)
+        }
+
+        if count < 0 {
+            return
+        }
+
+        _debugPrintCQE(header + " CQE:s [\(cqes)] - ring flags are [\(ring.flags)]")
+        for i in 0..<count {
+            let c = cqes[i]!.pointee
+
+            let bitPattern = UInt(bitPattern:io_uring_cqe_get_data(cqes[i]))
+            let uringUserData = URingUserData(rawValue: UInt64(bitPattern))
+
+            _debugPrintCQE("\(i) = fd[\(uringUserData.fileDescriptor)] eventType[\(String(describing:uringUserData.eventType))] registrationID[\(uringUserData.registrationID)] res [\(c.res)] flags [\(c.flags)]")
+        }
+        #endif
+    }
+
+    init() {
+        cqes = UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>.allocate(capacity: Int(cqeMaxCount))
+        cqes.initialize(repeating:&emptyCqe, count:Int(cqeMaxCount))
+    }
+
+    deinit {
+        cqes.deallocate()
+    }
+
+    internal func io_uring_queue_init() throws -> () {
+        if (CNIOLinux.io_uring_queue_init(ringEntries, &ring, 0 ) != 0)
+         {
+             throw UringError.uringSetupFailure
+         }
+
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+
+    internal func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux.io_uring_queue_exit(&ring)
+    }
+
+    static internal func io_uring_use_multishot_poll() -> Bool {
```

Comment:
Any reason this isn't a computed property?

Reply:
Same, changing.